core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -1967,39 +1967,138 @@ private[spark] object Utils extends Logging {
+        distinct.filterNot(_ == startPort)
+      val maxRetries = Math.min(portMaxRetries(conf), filteredPorts.size)
+
+      val availPorts = {
+        if (rand) {
+          scala.util.Random.shuffle(filteredPorts.take(maxRetries)).toArray
+        } else {
+          filteredPorts.sorted.toArray
+        }
+      }
+
+      val tryPort = (x: Int) => availPorts(x).toInt
+
+      retryPort(tryPort, maxRetries)
+    }
+
     require(startPort == 0 || (1024 <= startPort && startPort < 65536),
       "startPort should be between 1024 and 65535 (inclusive), or 0 for a random free port.")
 
-    val serviceString = if (serviceName.isEmpty) "" else s" '$serviceName'"
-    val maxRetries = portMaxRetries(conf)
-    for (offset <- 0 to maxRetries) {
+    if (hasPortRangeRestriction(conf)) {
+
+      if (startPort != 0) {
+
+        val (service, port) = startOnce(startPort)
+
+        if (port != -1) {
+          (service.get, port)
+        } else {
+          // try other
+          startFromAvailable()
+        }
+      } else {
+        // try random
+        startFromAvailable(true)
+      }
+
+    } else {
+
+      val maxRetries = portMaxRetries(conf)
+
       // Do not increment port if startPort is 0, which is treated as a special port
-      val tryPort = if (startPort == 0) {
+      val tryPort = (x: Int) => if (startPort == 0) {
         startPort
       } else {
         // If the new port wraps around, do not try a privilege port
-        ((startPort + offset - 1024) % (65536 - 1024)) + 1024
-      }
-      try {
-        val (service, port) = startService(tryPort)
-        logInfo(s"Successfully started service$serviceString on port $port.")
-        return (service, port)
-      } catch {
-        case e: Exception if isBindCollision(e) =>
-          if (offset >= maxRetries) {
-            val exceptionMessage =
-              s"${e.getMessage}: Service$serviceString failed after $maxRetries retries!"
-            val exception = new BindException(exceptionMessage)
-            // restore original stack trace
-            exception.setStackTrace(e.getStackTrace)
-            throw exception
-          }
-          logWarning(s"Service$serviceString could not bind on port $tryPort. " +
-            s"Attempting port ${tryPort + 1}.")
+        ((startPort + x - 1024) % (65536 - 1024)) + 1024
       }
+
+      retryPort(tryPort, maxRetries)
     }
-    // Should never happen
-    throw new SparkException(s"Failed to start service$serviceString on port $startPort")
+  }
+
+  /**
+   * Checks if there is any port restriction (currently used only with mesos)
+   * @param conf the SparkConfig to use
+   * @return true if port range restriction holds false otherwise
+   */
+  def hasPortRangeRestriction(conf: SparkConf): Boolean = {
+    // Caution: the property is for internal use only
+    conf.getOption("spark.mesos.executor.port.ranges").isDefined
+  }
+
+  /**
+   * Gets the allowed port ranges to use (currently  used only with mesos)
```

Comment:
if these are only used in Mesos they don't belong here. We should put them in `MesosSchedulerUtils` or something